src/messenger/messenger.cpp

```diff
@@ -0,0 +1,256 @@
+ *      sock.setsockopt(zmq.IDENTITY, b'snek')        # Set the name of our Python program,
+ *      sock.connect('ipc:///tmp/noisepage-ipc0')     # Connect to NoisePage on the same machine over IPC.
+ *      sock.send(b'', flags=zmq.SNDMORE)             # Start building a message. This is the empty delimiter packet.
+ *      s.send(b'PThis is the message payload.')      # Finish the message and send it. P is a prefix for Print.
+ *
+ */
+
+namespace terrier::messenger {
+
+/** An abstraction around ZeroMQ messages which explicitly have the sender specified. */
+class ZmqMessage {
+ public:
+  /** The routing ID of the message sender. */
+  std::string identity_;
+  /** The payload in the message. */
+  std::string payload_;
+};
+
+}  // namespace terrier::messenger
+
+namespace terrier {
+
+/**
+ * Useful ZeroMQ utility functions implemented in a naive manner. Most functions have wasteful copies.
+ * If perf indicates that these functions are a bottleneck, switch to the zero-copy messages of ZeroMQ.
+ */
+class ZmqUtil {
+ public:
+  /** ZmqUtil is a static utility class that should not be instantiated. */
+  ZmqUtil() = delete;
+
+  /** The maximum length of a routing ID. Specified by ZeroMQ. */
+  static constexpr int MAX_ROUTING_ID_LEN = 255;
+
+  /** @return The routing ID of the socket. */
+  static std::string GetRoutingId(common::ManagedPointer<zmq::socket_t> socket) {
+    char buf[MAX_ROUTING_ID_LEN];
+    size_t routing_id_len = MAX_ROUTING_ID_LEN;
+    socket->getsockopt(ZMQ_ROUTING_ID, &buf, &routing_id_len);
+    return std::string(buf, routing_id_len);
+  }
+
+  /** @return The next string to be read off the socket. */
+  static std::string Recv(common::ManagedPointer<zmq::socket_t> socket, zmq::recv_flags flags) {
+    zmq::message_t message;
+    auto received = socket->recv(message, flags);
+    if (!received.has_value()) {
+      throw MESSENGER_EXCEPTION(fmt::format("Unable to receive on socket: {}", ZmqUtil::GetRoutingId(socket)));
+    }
+    return std::string(static_cast<char *>(message.data()), received.value());
+  }
+
+  /** @return The next ZmqMessage (identity and payload) read off the socket. */
+  static messenger::ZmqMessage RecvMsg(common::ManagedPointer<zmq::socket_t> socket) {
+    std::string identity = Recv(socket, zmq::recv_flags::none);
+    std::string delimiter = Recv(socket, zmq::recv_flags::none);
+    std::string payload = Recv(socket, zmq::recv_flags::none);
+
+    return messenger::ZmqMessage{identity, payload};
+  }
+
+  /** @return Send the specified ZmqMessage (identity and payload) over the socket. */
+  static void SendMsg(common::ManagedPointer<zmq::socket_t> socket, const messenger::ZmqMessage &msg) {
+    zmq::message_t identity_msg(msg.identity_.data(), msg.identity_.size());
+    zmq::message_t delimiter_msg("", 0);
+    zmq::message_t payload_msg(msg.payload_.data(), msg.payload_.size());
+    bool ok = true;
+
```

Comment:
I actually managed to make it work without deleting the send `identity_msg`

Reply:
oh, interesting. That's weird. I wonder if that means the python API is adding it automatically but the C++ api does not.

Reply:
It turns out that ROUTER and DEALER sockets behave differently. I will update this information inside the slides.